Fix typing errors in set monster level action

ultidonki opened the pull request:
Firstly, we need to add a len() around the player.monsters list, as you can't compare a list and 0 and it was giving an error when the action was used. 
(I guess you could replace it with:
`if not self.session.player.monsters:`
I can change it if that's more Python-y? )

Secondly, add typing around the params, as you're allowed to pass None for both params, eg the action:
set_player_monster ,
Will set all player's monsters to level 1. 
Previously this was giving errors if you tried to pass None as a param.

Commit:
Fix typing errors in set monster level action

xirsoi commented:
Why the heck would you wanna set all the player's monsters to level 1?
I question the validity of `set_monster ,`

ultidonki commented:
Lol, I agree, I thought it was weird. I've looked up the original pull request and there's not much explanation there either. 

However, I can think of a possible use-case:
If you want to set all player monster's to level 50, you need to do:
set_monster_level ,
set_monster_level ,49
To first set all levels to 1, and then bump up all levels to 50. 

As far as I know no-one's using this, but it could be useful to someone in the future, maybe.

bitcraft commented on tuxemon/event/actions/set_monster_level.py:
```diff
@@ -21,12 +21,12 @@
 class SetMonsterLevelActionParameters(NamedTuple):
-    slot: int
-    level: int
+    slot: Union[int, None]
```
I think this is better typed as `Optional[int]`

vnmabus replied:
Usually yes, but here the type is being used to parse the parameter by trying to cast to each type sequentially in case of a Union (so the order matters). Maybe support for Optional could also be added, but for now, it is not.

bitcraft replied:
So this is fine as is?

vnmabus replied:
It should be, yes. We could try to accept also `Optional` in the parsing function, in a separate PR, and change every occurrence of `Union[whatever, None]` in event parameters. It is very low priority, though.

bitcraft replied:
ty for walking me through that vnmabus

ultidonki replied:
I've done some tests with Optional[int] based on this convo - in this action, the behaviour before and after was the same, both for valid params (int and none) and invalid params (float and str - they were accepted without raising an error in both the Union and Optional cases). 

I can make the change here if you want, as it seems the same, but I'll raise an issue seperately to track changing all the others.

vnmabus replied:
It seems that `Optional[whatever]` is translated at runtime to `Union[whatever, None]`. That is why it works. In our case, the order in which each type appears matters, as each type is tried sequentially, so we have to be 100% sure that future versions of Python do not change the order, and `None` is still the last parameter. If that is the case we could change them safely.

ultidonki replied:
I think you're right - I've looked at that code now (cast_values in tools.py) and if Optional was implemented in any other way, that code would fail. I don't think we want to rely on a quirk of the python implementation, so I'll keep it as Union and close the other issue I raised. Thanks for checking that!

vnmabus replied:
The official documentation says literally "Optional[X] is equivalent to Union[X, None]". I think that is unlikely that the behaviour changes in future versions, so if you want to use optional instead, it is ok for me.

ultidonki replied:
I'm happy either way - Union[X, None] reminds users that they need to handle both the X and None cases, which is good. But if Optional is used everywhere else in the code base then it might be better to be consistent. So I don't mind.

It's good to know that the choice is just a style choice, and both will have the same effect behind the scenes, I think that's the important thing.

bitcraft commented:
> if not self.session.player.monsters:

i would prefer it written like this.  the previous way of `if not self.session.player.monsters > 0` is kinda confusing.


> Will set all player's monsters to level 1.

i kinda agree with xirsoi here.  leaving out the param and then setting all the monster levels seems like unexpected behavior.  thinking about it...  i don't think any of the params should be optional, as any default there is kinda hard to reason with.  explicit is better than implicit, especially since the scripts are written and read by non-programmers and should be easy to understand without knowing too much.

Commit:
Easier to read expression in set_monster_level action

Commit:
Make set_monster_level keep new level between 1 and MAX_LEVEL

Commit:
Don't allow level param in set_monster_level action to be blank

Commit:
Add automated tests for set_level in monster class

ultidonki commented:
OK, I've added a couple of things to this PR - though I can break them into seperate PR's if that helps.

- I've changed the complicated expression to this one now:
`if not self.session.player.monsters:` 

- I've removed the ability to pass a blank parameter for 'level', so it no longer resets the monster's level to 1. (It turns out you can do this anyway by passing in negative numbers, eg. -999 will clamp it to 1, always.) Neither of the 2 official campaigns were using this functionality anyway, so nothing needs to be changed in those mods. 

-  I noticed this event actually allows you to go above MAX_LEVEL, so I pushed some logic into monster.py to clamp the new level between 1 and MAX_LEVEL, and made a call to that function from this action instead. 

- I also added some unit tests to the set_level function in monster.py just to make sure it always clamps between 1 and MAX_LEVEL now.

bitcraft commented:
Thank you!
